In [47]:
import pandas as pd

In [48]:
#Loading the dataset from excel to dataframe 
data = pd.read_excel('OnlineRetail.xlsx') 

In [49]:
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [50]:
#As each transaction is having a unique Invoice number, its count gives the total number of entries
data.InvoiceNo.count()

541909

There are 541909 entries in the dataframe, however the data cleaning operation is to be performed for R_F_M analysis

### Removing NaN entries from the dataset

In [51]:
data.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [52]:
data[data.CustomerID.isna()].head()

InvoiceNo StockCode                      Description  Quantity  \
622     536414     22139                              NaN        56   
1443    536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444    536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445    536544     21786               POLKADOT RAIN HAT          4   
1446    536544     21787            RAIN PONCHO RETROSPOT         2   

             InvoiceDate  UnitPrice  CustomerID         Country  
622  2010-12-01 11:52:00       0.00         NaN  United Kingdom  
1443 2010-12-01 14:32:00       2.51         NaN  United Kingdom  
1444 2010-12-01 14:32:00       2.51         NaN  United Kingdom  
1445 2010-12-01 14:32:00       0.85         NaN  United Kingdom  
1446 2010-12-01 14:32:00       1.66         NaN  United Kingdom

dropping rows where customerID is NaN, as it is an important column for R_F_M analysis

In [53]:
data.dropna(subset=['CustomerID'],inplace=True)

In [54]:
sum(data.CustomerID.isna())

0

In [55]:
#description column is not required for analysis
data1 = data.drop(labels='Description',axis = 1)

In [56]:
data1.head()

InvoiceNo StockCode  Quantity         InvoiceDate  UnitPrice  CustomerID  \
0    536365    85123A         6 2010-12-01 08:26:00       2.55     17850.0   
1    536365     71053         6 2010-12-01 08:26:00       3.39     17850.0   
2    536365    84406B         8 2010-12-01 08:26:00       2.75     17850.0   
3    536365    84029G         6 2010-12-01 08:26:00       3.39     17850.0   
4    536365    84029E         6 2010-12-01 08:26:00       3.39     17850.0   

          Country  
0  United Kingdom  
1  United Kingdom  
2  United Kingdom  
3  United Kingdom  
4  United Kingdom

### Checking for negative values.

In [57]:
#quantity cannot be negative
sum(data1['Quantity']<0)

8905

In [62]:
data1[data1['Quantity']<0].head()

InvoiceNo StockCode  Quantity         InvoiceDate  UnitPrice  CustomerID  \
141   C536379         D        -1 2010-12-01 09:41:00      27.50     14527.0   
154   C536383    35004C        -1 2010-12-01 09:49:00       4.65     15311.0   
235   C536391     22556       -12 2010-12-01 10:24:00       1.65     17548.0   
236   C536391     21984       -24 2010-12-01 10:24:00       0.29     17548.0   
237   C536391     21983       -24 2010-12-01 10:24:00       0.29     17548.0   

            Country  
141  United Kingdom  
154  United Kingdom  
235  United Kingdom  
236  United Kingdom  
237  United Kingdom

there are negative values on Quantity variable which is caused by the refund invoices (Invoices containing the letter "C")

In [63]:
index_no = data1[data1['Quantity']<0].index

In [64]:
data1 = data1.drop(index_no,axis=0)

In [65]:
sum(data1['Quantity']<0)

0

##### Negative quantities values removed

In [66]:
sum(data1.StockCode.isna())

0

In [67]:
sum(data1.InvoiceNo.isna())

0

#### All the relevant NaN values and where Quantity is less than zero are removed from the dataset. Since description is not an important parameter in our analysis, its NaN values can be ignored 

### Removing Duplicates 

In [68]:
data1.Country.unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Finland',
       'Austria', 'Greece', 'Singapore', 'Lebanon',
       'United Arab Emirates', 'Israel', 'Saudi Arabia', 'Czech Republic',
       'Canada', 'Unspecified', 'Brazil', 'USA', 'European Community',
       'Bahrain', 'Malta', 'RSA'], dtype=object)

In [69]:
len(data1.Country.unique())

37

There are 37 unique countries in the dataset

In [70]:
sum(data1.duplicated())

5194

There are 5194 duplicate entries in the dataset which are to be removed 

In [71]:
data1 = data1.drop_duplicates()

In [72]:
sum(data1.duplicated())

0

#### Thus all the duplicated from the dataset are removed

the names of the countries with the total values:

In [73]:
data1['Country'].value_counts()

United Kingdom          349225
Germany                   9027
France                    8327
EIRE                      7228
Spain                     2480
Netherlands               2363
Belgium                   2031
Switzerland               1842
Portugal                  1453
Australia                 1184
Norway                    1072
Italy                      758
Channel Islands            747
Finland                    685
Cyprus                     603
Sweden                     450
Austria                    398
Denmark                    380
Poland                     330
Japan                      321
Israel                     245
Unspecified                241
Singapore                  222
Iceland                    182
USA                        179
Canada                     151
Greece                     145
Malta                      112
United Arab Emirates        68
European Community          60
RSA                         58
Lebanon                     45
Lithuani

### Performing R_F_M analysis

In [75]:
data1['TotalPrice'] = data1['Quantity']*data1['UnitPrice']

In [76]:
data1.head()

InvoiceNo StockCode  Quantity         InvoiceDate  UnitPrice  CustomerID  \
0    536365    85123A         6 2010-12-01 08:26:00       2.55     17850.0   
1    536365     71053         6 2010-12-01 08:26:00       3.39     17850.0   
2    536365    84406B         8 2010-12-01 08:26:00       2.75     17850.0   
3    536365    84029G         6 2010-12-01 08:26:00       3.39     17850.0   
4    536365    84029E         6 2010-12-01 08:26:00       3.39     17850.0   

          Country  TotalPrice  
0  United Kingdom       15.30  
1  United Kingdom       20.34  
2  United Kingdom       22.00  
3  United Kingdom       20.34  
4  United Kingdom       20.34

In [77]:
#Checking the last date of purchase and assigning "today's date" as 1 day after the last date of purchase 
#to make sure that none of the Recency values become zero
data1['InvoiceDate'].max()

Timestamp('2011-12-09 12:50:00')

In [79]:
import datetime as dt
today = dt.datetime(2011,12,10)
data1['InvoiceDate'] = pd.to_datetime(data1['InvoiceDate'])

### Created a new df called Table_rfm and renaming its column in order to calculate Recency, Frequency and Monetary values.
### df is grouped by customers and:
### Recency is the number of days between today's date that we select and the last purchase date of this customer 
### Frequency is the number of unique invoices of this customer 
### Customer's Monetary  is the sum of TotalPrice

In [80]:
Table_rfm = data1.groupby('CustomerID').agg({'InvoiceDate': lambda x: (today - x.max()).days, 
                                             'InvoiceNo': lambda x: len(x), 'TotalPrice': lambda x: x.sum()})
Table_rfm['InvoiceDate'] = Table_rfm['InvoiceDate'].astype(int)

#renaming the columns
Table_rfm.rename(columns={'InvoiceDate': 'Recency', 
                         'InvoiceNo': 'Frequency', 
                         'TotalPrice': 'Monetary_value'}, inplace=True)

In [82]:
Table_rfm.head()

Recency  Frequency  Monetary_value
CustomerID                                    
12346.0         325          1        77183.60
12347.0           2        182         4310.00
12348.0          75         31         1797.24
12349.0          18         73         1757.55
12350.0         310         17          334.40

In [83]:
# To check if there are any zeros in the RFM

Table_rfm.describe([0.01, 0.05, 0.10, 0.25, 0.50, 0.75, 0.90, 0.95, 0.99]).T

count         mean          std  min      1%       5%  \
Recency         4339.0    92.041484   100.007757  0.0   1.000    2.000   
Frequency       4339.0    90.511639   225.515337  1.0   1.000    4.000   
Monetary_value  4339.0  2048.088740  8983.700739  0.0  52.076  111.647   

                    10%      25%     50%       75%       90%       95%  \
Recency           5.000   17.000   50.00   141.500   262.200   311.000   
Frequency         7.000   17.000   41.00    98.000   202.200   308.000   
Monetary_value  155.146  306.455  668.56  1660.315  3640.784  5787.243   

                       99%        max  
Recency           368.6200     373.00  
Frequency         655.6200    7676.00  
Monetary_value  19780.2672  280206.02

# RFM Scores
### The highest score should be given to the lower number of Recency as the min number of Recency metric means that this customer has just purchased something.
### The highest score should be given to the highest Frequency and Monetary values as the max number of Frequency and Monetary metrics means that the customer is purchasing frequently and spending more money.

In [85]:
quantiles = Table_rfm.quantile(q=[0.25,0.5,0.75])
quantiles = quantiles.to_dict()

In [86]:
# creating segmented_RFM table

seg_rfm = Table_rfm

In [87]:
# calculated the lowest recency, highest frequency and monetary amounts of the best customers

def R_Score(x,p,d):
    if x <= d[p][0.25]:
        return 1
    elif x <= d[p][0.50]:
        return 2
    elif x <= d[p][0.75]: 
        return 3
    else:
        return 4
    
def FM_Score(x,p,d):
    if x <= d[p][0.25]:
        return 4
    elif x <= d[p][0.50]:
        return 3
    elif x <= d[p][0.75]: 
        return 2
    else:
        return 1

In [89]:
#using quartiles to split the metrics into segments and adding segmented numbers into newly created 
# segmented_RFM table

seg_rfm['r_quartile'] = seg_rfm['Recency'].apply(R_Score, args=('Recency',quantiles,))
seg_rfm['f_quartile'] = seg_rfm['Frequency'].apply(FM_Score, args=('Frequency',quantiles,))
seg_rfm['m_quartile'] = seg_rfm['Monetary_value'].apply(FM_Score, args=('Monetary_value',quantiles,))
seg_rfm.head()

Recency  Frequency  Monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
12346.0         325          1        77183.60           4           4   
12347.0           2        182         4310.00           1           1   
12348.0          75         31         1797.24           3           3   
12349.0          18         73         1757.55           2           2   
12350.0         310         17          334.40           4           4   

            m_quartile  
CustomerID              
12346.0              1  
12347.0              1  
12348.0              1  
12349.0              1  
12350.0              3

In [90]:
# string concatination: Forming a new column to combine RMF_Score

seg_rfm['RFM_Score'] = seg_rfm.r_quartile.map(str) + seg_rfm.f_quartile.map(str) + seg_rfm.m_quartile.map(str)
seg_rfm.head()

Recency  Frequency  Monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
12346.0         325          1        77183.60           4           4   
12347.0           2        182         4310.00           1           1   
12348.0          75         31         1797.24           3           3   
12349.0          18         73         1757.55           2           2   
12350.0         310         17          334.40           4           4   

            m_quartile RFM_Score  
CustomerID                        
12346.0              1       441  
12347.0              1       111  
12348.0              1       331  
12349.0              1       221  
12350.0              3       443

In [94]:
#top 10 customers with the highest RFM_scores sorted based on Recency

seg_rfm[seg_rfm['RFM_Score']=='111'].sort_values('Recency', ascending=False).head(10)

Recency  Frequency  Monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
13268.0          17        433         3174.75           1           1   
14505.0          17        785         5859.43           1           1   
16655.0          17        258         3789.52           1           1   
16722.0          17        225         4730.81           1           1   
16633.0          17        152         2883.10           1           1   
12656.0          17        126         4051.17           1           1   
13488.0          17        422         8968.31           1           1   
18139.0          17        159         8438.34           1           1   
17659.0          17        158         2999.46           1           1   
15845.0          17        138         3711.77           1           1   

            m_quartile RFM_Score  
CustomerID                        
13268.0              1       111  
14505.0              1       111  
16655.0              1       111  
16722.0              1       111  
16633.0              1       111  
12656.0              1       111  
13488.0              1       111  
18139.0              1       111  
17659.0              1       111  
15845.0              1       111

In [92]:
#top 10 customers with the highest RFM_scores sorted based on Frequency

seg_rfm[seg_rfm['RFM_Score']=='111'].sort_values('Frequency', ascending=False).head(10)

Recency  Frequency  Monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
17841.0           1       7676        40519.84           1           1   
14911.0           1       5672       143711.17           1           1   
14096.0           4       5111        65164.79           1           1   
12748.0           0       4413        33053.19           1           1   
14606.0           1       2677        12076.15           1           1   
15311.0           0       2366        60632.75           1           1   
14646.0           1       2080       280206.02           1           1   
13089.0           2       1814        58762.08           1           1   
13263.0           1       1667         7436.93           1           1   
14298.0           8       1637        51527.30           1           1   

            m_quartile RFM_Score  
CustomerID                        
17841.0              1       111  
14911.0              1       111  
14096.0              1       111  
12748.0              1       111  
14606.0              1       111  
15311.0              1       111  
14646.0              1       111  
13089.0              1       111  
13263.0              1       111  
14298.0              1       111

In [91]:
#top 10 customers with the highest RFM_scores sorted based on Monetary_value

seg_rfm[seg_rfm['RFM_Score']=='111'].sort_values('Monetary_value', ascending=False).head(10)

Recency  Frequency  Monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
14646.0           1       2080       280206.02           1           1   
18102.0           0        431       259657.30           1           1   
17450.0           8        336       194390.79           1           1   
14911.0           1       5672       143711.17           1           1   
14156.0           9       1395       117210.08           1           1   
17511.0           2        963        91062.38           1           1   
16684.0           4        277        66653.56           1           1   
14096.0           4       5111        65164.79           1           1   
13694.0           3        568        65039.62           1           1   
15311.0           0       2366        60632.75           1           1   

            m_quartile RFM_Score  
CustomerID                        
14646.0              1       111  
18102.0              1       111  
17450.0              1       111  
14911.0              1       111  
14156.0              1       111  
17511.0              1       111  
16684.0              1       111  
14096.0              1       111  
13694.0              1       111  
15311.0              1       111

In [100]:
# Naming the RFM score

seg_map = {
    r'[1-2][1-2]': 'Hibernating',
    r'[1-2][3-4]': 'At Risk',
    r'[1-2]5': 'Can\'t Loose',
    r'3[1-2]': 'About to Sleep',
    r'33': 'Need Attention',
    r'[3-4][4-5]': 'Loyal Customers',
    r'41': 'Promising',
    r'51': 'New Customers',
    r'[4-5][2-3]': 'Potential Loyalists',
    r'5[4-5]': 'Champions'
}`

In [106]:
# the customer who is often purchasing and recently purchased have high RFM_Score
# Therefore Recency score and Frequency score is used for customer segmentation

seg_rfm['Segment'] = seg_rfm['r_quartile'].astype(str) + seg_rfm['f_quartile'].astype(str)
seg_rfm.head()

Recency  Frequency  Monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
12346.0         325          1        77183.60           4           4   
12347.0           2        182         4310.00           1           1   
12348.0          75         31         1797.24           3           3   
12349.0          18         73         1757.55           2           2   
12350.0         310         17          334.40           4           4   

            m_quartile RFM_Score Segment  
CustomerID                                
12346.0              1       441      44  
12347.0              1       111      11  
12348.0              1       331      33  
12349.0              1       221      22  
12350.0              3       443      44

In [107]:
# converting the metrics into the category names

seg_rfm['Segment'] = seg_rfm['Segment'].replace(seg_map, regex=True)
seg_rfm.head()

Recency  Frequency  Monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
12346.0         325          1        77183.60           4           4   
12347.0           2        182         4310.00           1           1   
12348.0          75         31         1797.24           3           3   
12349.0          18         73         1757.55           2           2   
12350.0         310         17          334.40           4           4   

            m_quartile RFM_Score          Segment  
CustomerID                                         
12346.0              1       441  Loyal Customers  
12347.0              1       111      Hibernating  
12348.0              1       331   Need Attention  
12349.0              1       221      Hibernating  
12350.0              3       443  Loyal Customers

In [109]:
# displaying the number of customers that fall into each category

seg_rfm['Segment'].value_counts()

Hibernating            1485
Loyal Customers         824
At Risk                 692
Potential Loyalists     515
About to Sleep          451
Need Attention          326
Promising                46
Name: Segment, dtype: int64

In [110]:
seg_rfm[["Segment", "Recency", "Frequency", "Monetary_value"]].groupby("Segment").agg(["mean", "count"])

Recency         Frequency       Monetary_value      
                           mean count        mean count           mean count
Segment                                                                     
About to Sleep        80.807095   451   96.771619   451    1660.370266   451
At Risk               23.517341   692   20.744220   692     870.338324   692
Hibernating           17.437037  1485  196.020202  1485    4356.186277  1485
Loyal Customers      195.292476   824    8.911408   824     526.928448   824
Need Attention        85.257669   326   28.457055   326     708.662917   326
Potential Loyalists  237.601942   515   38.365049   515     605.852466   515
Promising            210.347826    46  157.891304    46    1943.307413    46